src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift

```diff
@@ -1493,6 +1511,83 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
 		
 		push(exposureSubmissionSuccessViewController)
 	}
+
+	private func restoreAndStartWithTest(from recycleBinItem: RecycleBinItem) {
+		model.recycleBin.restore(recycleBinItem)
+
+		switch recycleBinItem.item {
+		case .certificate:
+			Log.info("restoreAndShowTest only restores tests")
+			return
+		case .userCoronaTest(let coronaTest):
+			start(with: coronaTest.type)
+		case .familyMemberCoronaTest(let coronaTest):
+			let familyMemberTestResultViewController = createTestResultScreen(for: coronaTest)
+			start(with: familyMemberTestResultViewController)
+		}
+	}
+
+	private func showTestRestoredFromBinAlert(
+		recycleBinItem: RecycleBinItem
+	) {
+		let alert = UIAlertController(
+			title: AppStrings.UniversalQRScanner.testRestoredFromBinAlertTitle,
+			message: AppStrings.UniversalQRScanner.testRestoredFromBinAlertMessage,
+			preferredStyle: .alert
+		)
+		alert.addAction(
+			UIAlertAction(
+				title: AppStrings.Common.alertActionOk,
+				style: .default,
+				handler: { _ in
+					switch self.model.recycleBin.canRestore(recycleBinItem) {
+					case .success:
+						self.restoreAndStartWithTest(from: recycleBinItem)
+					case .failure(.testError(.testTypeAlreadyRegistered)):
+						self.showTestOverwriteNotice(recycleBinItem: recycleBinItem)
+					}
+				}
+			)
+		)
+
+		parentViewController?.present(alert, animated: true)
```

Comment:
Dispatch on main queue needed?

Reply:
I tried it out, there seems to be no queue change for the dismiss closure, so no dispatch seems to be needed